Fixed mean() and moment() on sparse arrays

pentschev opened the pull request:
For proper creation of sparse auxiliary arrays, a dispatcher for ones_like has
been added. However, sparse arrays created from ones_like cause sparse
reductions to fail, as it would result in a dense array. To fix that, the arrays
have first to be converted to dense before, which ends up requiring a new
dispatcher for that purpose, and for already dense arrays must be bypassed, that
is here implemented as numpy.ndarray.view() to simply return a shallow copy of
the array.

This commit fixes https://github.com/dask/dask/issues/4523 and adds the tests
suggested in that issue.

@mrocklin 

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

pentschev commented:
@mrocklin please let me know what you think of this. I'm not particularly happy with the need of the `todense()` dispatcher I added, but it doesn't seem like a terrible solution either.

mrocklin commented:
I think that we shouldn't ever have to create the original ones array.  Instead we should figure out what the output looks like and create that.  

```python
In [2]: from dask.array.reductions import numel

In [3]: import numpy as np

In [4]: x = np.ones((2, 3, 4))

In [5]: numel(x, axis=(2,), keepdims=False)
Out[5]:
array([[4., 4., 4.],
       [4., 4., 4.]])

In [6]: numel(x, axis=(2,), keepdims=False).shape
Out[6]: (2, 3)

In [7]: np.ones(shape=x.shape[:2]) * x.shape[2]
Out[7]:
array([[4., 4., 4.],
       [4., 4., 4.]])
```

This gets a little bit tricky for complex values of `axis=` and `keepdims=`, but avoids us having to create a large dense array on the CPU.  We only have to create the much smaller reduced array.

mrocklin commented:
And actually, going even further, we may not need to track numel at all.  I think that we may be able to just track a single integer throughout all of the cases where numel is used.  If so, this would be preferable.

So in general I think that instead of improving numel we should take a hard look at how it is used today, and see if we can come up with a better algorithm (if you're up for that).

pentschev commented:
I agree that not creating a the original ones array would be a better solution. However, if I'm not missing something, we still can't create a `ones()` array of arbitrary shape here, because the backend of `x` is unknown (i.e., think of CuPy). Calculating the final array is tricky but doable, but we still depend on the `ones_like()` for that case, until we have a solution on creating arrays with the appropriate backend.

mrocklin commented:
It may be that we never need to create the array of ones at all.  It may be that we're passing way more information than we need to.   I think that we only need the single value stored in the array, along with the shape of the array (which happens to be the same as the shape of the other arrays currently passed in the same dictionary).

pentschev commented:
@mrocklin I changed the `numel()` function not to create an array of ones, but compute them with the input parameters. Exceptions are masked arrays and the `nannumel()` that can't really be computed without going through the `x` array anyway.

This still doesn't solve returning only the value and shape, but it's a start.

Also, I'm sure the `numel()` implementation is not in its best "Pythonic" form, feel free to criticize it.

pentschev commented:
And I almost forgot, but the `todense()` has to remain in `moment_chunk()` for now, since sparse doesn't implement `sum()`.

mrocklin commented:
This looks great to me.  I'm glad to see this come together.  

There appears to be a failure in the Python 2.7 build, it looks like behavior might have changed in one of the relevant libraries.

pentschev commented on dask/array/tests/test_reductions.py:
```diff
@@ -15,6 +17,23 @@ def assert_eq(a, b):
+    x = np.ones((2, 3, 4))
+
+    for length in range(x.ndim):
+        for sub in itertools.combinations([x for x in range(x.ndim)], length):
```
So it turns out that Python 2.7 was overwriting the `x` array created before with the `x` from this list comprehension. Do you know which one is the "correct" behavior, @mrocklin? I'm supposing Python standard defines this somewhere, and perhaps this changed in Python 3.

mrocklin replied:
Yeah, I think that Python 3 handled scoping issues like this a bit better.  Temporary variables within a comprehension no longer overwrite variables in the outer scope.

pentschev replied:
Thanks for the info @mrocklin.

mrocklin commented:
This looks great to me.  Thanks @pentschev !  Merging.

mrocklin commented:
Actually, can we safely remove the `ones_like_lookup` now?

pentschev commented:
> Actually, can we safely remove the `ones_like_lookup` now?

Yes, I thought I would leave it as it could be useful, but probably creating a `ones()` sparse matrix isn't that useful, I'll remove that.

mrocklin commented:
So, I think that we can avoid implementing the `todense` dispatch (I'd really like to avoid having to implement these if we can) if `sparse` supports `sum` on arrays with fill values.  I've reported this upstream at https://github.com/pydata/sparse/issues/237

mrocklin commented:
I hope you don't mind, but I pushed a commit to your branch that removes the `todense_lookup`.  I think that long term it will be better to fix this upstream in `sparse`.  My guess is that we can get other folks to handle that though.

pentschev commented:
> So, I think that we can avoid implementing the todense dispatch (I'd really like to avoid having to implement these if we can) if sparse supports sum on arrays with fill values. I've reported this upstream at pydata/sparse#237

That's also what I had in mind.

> I hope you don't mind, but I pushed a commit to your branch that removes the todense_lookup. I think that long term it will be better to fix this upstream in sparse. My guess is that we can get other folks to handle that though.

No worries, I see you marked it xfail for now. Hopefully when `sum()` is supported, this will automatically work.

hameerabbasi commented:
You can remove the `xfail`, as soon as pydata/sparse#238 is merged, all will be fine. Tests are green, just waiting for CI to pass.

pentschev commented:
@hameerabbasi that was fast, thanks!

Didn't push yet, but now it fails when calling `np.stack()` on a sparse array, which means, no `__array_function__`. Should we also add it to sparse?

hameerabbasi commented:
That will have to wait for a day or so. Of course, you're free to make a PR and I'll review it.

pentschev commented:
I can send a PR, the question was more if that's something we should do, but I got my answer now. :)

pentschev commented on dask/array/reductions.py:
```diff
@@ -410,8 +409,7 @@ def moment_chunk(A, order=2, sum=chunk.sum, numel=numel, dtype='f8', **kwargs):
     diff = A - u
-    todense = todense_lookup.dispatch(type(diff))
-    xs = [sum(todense(A - u)**i, dtype=dtype, **kwargs) for i in range(2, order + 1)]
+    xs = [((A - u)**i).sum(dtype=dtype, **kwargs) for i in range(2, order + 1)]
```
@mrocklin this is the line that causes `nanvar` test to fail. Please note that `sum()` here is not Python's nor NumPy's `sum()`, but it's one of `moment_chunk()`'s arguments. I'm now wondering third things:

1. Should we rename the `sum` argument to something like `sum_func` to avoid confusion?
1. Is it possible there's a bug in the `sum()` function in `dask/array/reductions.py`? I don't think they should give us different results anyway.
1. Did we accidentally changed behavior in some reduction functions when `sum()` was replaced by `_concatenate2().sum()`?

pentschev commented:
As @hameerabbasi pointed out, we can now remove the xfail, but for that we'll need to use upstream sparse or wait for a new release. How should we handle that?

hameerabbasi commented:
I may be wrong, but Dask tests against Sparse master.

pentschev commented:
@hameerabbasi I've asked Travis to confirm that. :)

mrocklin commented:
Looks like we're good here?  OK to merge?

hameerabbasi commented on dask/array/tests/test_sparse.py:
```diff
@@ -34,8 +34,6 @@
     lambda x: da.transpose(x, (1, 2, 0)),
     lambda x: x.sum(),
     lambda x: x.mean(),
-    pytest.param(lambda x: x.std(),
```
It seems you removed the entire test instead of the xfail.

pentschev replied:
Oh damn, thanks for noticing that, let me add it back.

hameerabbasi commented on dask/array/reductions.py:
```diff
@@ -408,8 +408,7 @@ def moment_chunk(A, order=2, sum=chunk.sum, numel=numel, dtype='f8', **kwargs):
     u = total / n
-    diff = A - u
-    xs = [((A - u)**i).sum(dtype=dtype, **kwargs) for i in range(2, order + 1)]
+    xs = [sum((A - u)**i, dtype=dtype, **kwargs) for i in range(2, order + 1)]
```
`sum(...)` is much slower than `(...).sum()` for cases where the former works.

hameerabbasi replied:
Unless you're using `np.sum`...

pentschev replied:
This code is misleading, this should probably change at some point. Check my comment: https://github.com/dask/dask/pull/4525#discussion_r264689869.

pentschev commented:
So it turns out that Dask is not using upstream Sparse. Should I add the xfail back? @mrocklin I can either do that or we can try to use sparse upstream.

mrocklin commented:
I recommend adding a skipif decorator with a sparse version requirement

On Thu, Mar 14, 2019 at 2:13 PM Peter Andreas Entschev <
notifications@github.com> wrote:

> So it turns out that Dask is not using upstream Sparse. Should I add the
> xfail back? @mrocklin <https://github.com/mrocklin> I can either do that
> or we can try to use sparse upstream.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4525#issuecomment-473065391>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszMPvSyfFyD85x9hrtBtD9i-_iZARks5vWrt_gaJpZM4bP645>
> .
>

hameerabbasi commented:
https://github.com/dask/dask/blob/e5948244baacba775790ab1845ca059a37c1a7e1/continuous_integration/travis/install.sh#L78

Looks like it isn't, and it was dropped because of the Python 2 compatibility change. With NumPy and SciPy dropping Python 2 already in `master`, is it really worth it to keep it in Dask?

pentschev commented:
@mrocklin good idea.
@hameerabbasi which version should we expect to have `__array_function__` in?

hameerabbasi commented:
Non experimental in 1.17 at the earliest, which won't have Python 3.

pentschev commented:
Sorry, I meant Sparse. I have enabled the experimental flag for Dask builds with NumPy >= 1.16.

hameerabbasi commented:
I can make a release right now if it helps. :-)

pentschev commented:
That would be cool, I think it would allow us to test this properly with Sparse.

hameerabbasi commented:
Okay, release is up on PyPI, conda-forge soon to follow.

pentschev commented:
Nice. Is there a way to retrigger the build here @mrocklin?

hameerabbasi commented:
Just push a commit with the message "re-trigger CI" and no file changes.

pentschev commented:
I was hoping there was some way to avoid that, but let's do it then. :)

hameerabbasi commented:
Hmm. One test failed. You're probably missing a `.todense()` somewhere, though I don't know which function it is (the logs just show a generic `lambda` -- Not helpful).

pentschev commented:
It's on this line: https://github.com/dask/dask/blob/496fb5da063c012c4dfd44c28cf49406045fb2c5/dask/array/reductions.py#L459. 

However, I feel like @mrocklin will not allow a `.todense()` call in here, so maybe I'll have to check if it's possible to come up with a better solution.

hameerabbasi commented:
I might debug... Sparse might have regressed and returned a dense array where one would be sparse, I’ll look into it today.

pentschev commented:
@hameerabbasi I see. Thanks for that, and let me know if there's anything I can do to help you.

hameerabbasi commented:
@pentschev A reproducible case would be nice... I'm quite busy today and over the weekend so I may not be able to get to it, but that would certainly help.

pentschev commented:
@hameerabbasi posted a snippet in https://github.com/pydata/sparse/issues/242. I can also take a look, you think broadcast is most likely the issue?

hameerabbasi commented:
pydata/sparse#243 should fix this.

hameerabbasi commented:
I can make another release but I'd really like it if Dask tested against sparse master... Using `pytest.importorskip(...)` or other skipping methods.

pentschev commented:
Testing against sparse master would be nice, at least one of the builds. I can confirm https://github.com/pydata/sparse/pull/243 works locally, but for the build here to pass we would need another release or testing against master, I don't know if there's a simple way to do this for the latter.

hameerabbasi commented:
Oh, there is. Just replace the Travis-CI line I linked earlier with installing from source: https://sparse.pydata.org/en/latest/install.html

pentschev commented:
@hameerabbasi thanks for that, I didn't know you could install from a git repository with `pip`.

pentschev commented:
It still doesn't look like Travis is installing upstream Sparse, builds that should be using and should not be using upstream still xfail:

```python
XFAIL dask/array/tests/test_sparse.py::test_basic[<lambda>15]
  fixed in https://github.com/pydata/sparse/pull/243
XFAIL dask/array/tests/test_sparse.py::test_basic[<lambda>16]
  fixed in https://github.com/pydata/sparse/pull/243
```

@mrocklin is there a better way to debug why upstream sparse is not getting installed than me pushing commits every 30min to see if anything changes?

TomAugspurger commented:
It seems to have been installed: https://travis-ci.org/dask/dask/jobs/506806663#L1458

pentschev commented:
@TomAugspurger aha! I didn't see there was a collapsed part with versions there. It's indeed installed, thanks for pointing that out.

Unfortunately for me, that's bad. I can't seem to reproduce the xfail locally.

pentschev commented:
Alright, that's what was missing, I didn't actually enable `__array_function__` before, we totally missed that. Now it's been enabled and the sparse tests xpassed, as expected.

But now I see something really strange, the build claims to have passed but two tests failed: https://travis-ci.org/dask/dask/jobs/506916217#L1596

@mrocklin @TomAugspurger any ideas why it's reporting that it passed?

TomAugspurger commented:
Looks like lambdas 12 and 13 failed. Those are the two `np.fft` functions. Does NumPy dispatch those correctly?

TomAugspurger commented:
Oh, sorry, just saw your actual question. No idea why GitHub reported it passing...

TomAugspurger commented:
Ah, probably the addition of

```
export NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1
```

in `run_tests.sh` is making that script always exit with 1. Probably need a `set -e` or whatever, to stop the script when there's an error, so that Travis sees it and fails.

pentschev commented:
@TomAugspurger thanks for the hint, you were right about that export. That was anyway supposed to have been removed.

pentschev commented:
It took me quite some time to figure out what was the failure here.

1. We didn't really enable `__array_function__` testing when it was merged, that's why it always passed;
1. When calling `np.fft.fft()` and its derivatives, they get masked by `mkl_fft`, and they return `np.ndarray`'s instead of `da.Array`.

I believe this could be related to https://github.com/dask/dask/issues/3281, and maybe it would be interesting in the future to fix that, but for now this fix should suffice for the purpose of this PR.

@jakirkham FYI

pentschev commented:
This is now ready for review. Hopefully, we can get it merged this time.

jakirkham commented on .travis.yml:
```diff
@@ -47,10 +47,12 @@ jobs:
       - PYTHON=3.7
       - NUMPY=1.16.2
       - PANDAS>=0.24.1
+      - UPSTREAM_SPARSE=1
```
Thoughts on consolidating this with the `UPSTREAM_DEV` case below?

pentschev replied:
I purposely didn't add it there, so we can test both master _and_ current release.

jakirkham replied:
Sure that makes sense.

What I meant to ask (and failed to), is any thoughts on having the dev version of sparse installed alongside the other dev versions and making this case the stable release of sparse?

hameerabbasi replied:
I think the answer here is that `__array_function__` was very recently added to Sparse by @pentschev and we currently need upstream to test it. Though if it is a concern I can make another release within a few hours.

jakirkham replied:
Sure. Am just suggesting we move both of these (`UPSTREAM_SPARSE` and `*array_function`) into the next CI case, which is already using other nightlies (including one for NumPy).

Not pushing on releasing things or other complex changes. Just encouraging all the tests that require unreleased changes to live in the same test case.

Does that make sense? Please feel free to ask questions if not.

pentschev replied:
I understand your point @jakirkham, but I'd like these tests to be executed also during PRs, which is not the case for `UPSTREAM_DEV=1` case. I've been working on these issues very frequently, having theses tests outside of PRs will incur in missing such errors. I also think we should test `__array_function__` against NumPy 1.16.2, it should work there too, not just with upstream NumPy.

My idea is that we should do that later, once we're more confident with the stability of these new Dask, NumPy and Sparse features, but for now it's best to have more situations tested during PRs.

jakirkham commented on dask/array/reductions.py:
```diff
@@ -320,7 +320,30 @@ def nanmax(a, axis=None, keepdims=False, split_every=None, out=None):
+    if not isinstance(axis, tuple or list):
+        axis = [axis]
+    else:
+        axis = sorted(axis)
```
Just curious, why does it matter if it is sorted?

pentschev replied:
It doesn't. It might speedup search, but I now got rid of it in favor of readability.

jakirkham commented on dask/array/reductions.py:
```diff
@@ -320,7 +320,30 @@ def nanmax(a, axis=None, keepdims=False, split_every=None, out=None):
+    if keepdims is True:
+        new_shape = [shape[dim] if dim not in axis else 1 for dim in range(len(shape))]
+    else:
+        new_shape = [shape[dim] for dim in range(len(shape)) if dim not in axis]
```
These should be `tuple`s since they are shapes.

pentschev replied:
Fixed.

pentschev commented:
Changes applied, from my side it's ready for more review/merge.

pentschev commented:
@jakirkham thanks for review and suggestions.

jakirkham commented:
Thanks @pentschev. I think the only question is how we handle the CI point above. Though no strong thoughts on it personally. Otherwise LGTM.

pentschev commented:
Any more comments here, or is it ready for merge?

pentschev commented:
Argh, clicked close by accident. Hopefully there won't be any failures. :)

mrocklin commented:
@jakirkham can I ask you to make the final call on this?  I think that you're more familiar with it than I am.

hameerabbasi commented on dask/array/tests/test_reductions.py:
```diff
@@ -22,6 +22,10 @@ def assert_eq(a, b):
     x = np.ones((2, 3, 4))
 
+    for length in range(x.ndim):
+        assert_eq(da.reductions.numel(x, axis=0, keepdims=keepdims, dtype=dtype),
```
Are you sure this is right? `length` isn't used anywhere in this loop.

pentschev replied:
Right, thanks for noticing that. Did it in a hurry. :)

pentschev commented:
Failures are test_parquet, therefore, not related to this PR.

@jakirkham are we good to merge this?

pentschev commented:
Everything passes again.

pentschev commented:
Earlier today, @jakirkham and I discussed this PR offline, and I understood what was his intent regarding upstream libraries. I therefore did now what we agreed to try, which is to reenable the development build that uses upstream libraries.

jakirkham commented:
Thanks @pentschev! LGTM

@jcrist and @martindurant, thoughts?

jakirkham commented:
Am seeing a few errors coming from Sparse/Numba in the dev tests that look like this

```
tp = <class 'numba.errors.TypingError'>
value = TypingError('Failed in nopython mode pipeline (step: nopython frontend)\nFailed in nopython mode pipeline (step: nopyt...rt the error message\nand traceback, along with a minimal reproducer at:\nhttps://github.com/numba/numba/issues/new\n')
tb = None
    def reraise(tp, value, tb=None):
        if value is None:
            value = tp()
        if value.__traceback__ is not tb:
>           raise value.with_traceback(tb)
E           numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E           Failed in nopython mode pipeline (step: nopython frontend)
E           Invalid use of Function(<function searchsorted at 0x7f37444bb840>) with argument(s) of type(s): (array(int64, 1d, C), int64, side=Literal[str](left))
E            * parameterized
E           In definition 0:
E               TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E           Invalid use of Function(<function _searchsorted.<locals>.searchsorted_inner at 0x7f37048f27b8>) with argument(s) of type(s): (array(int64, 1d, C), int64)
E            * parameterized
E           In definition 0:
E               TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E           Invalid use of Function(<ufunc 'isnan'>) with argument(s) of type(s): (int64)
E            * parameterized
E           In definition 0:
E               TypingError: ufunc 'isnan' using the loop 'l->?' not supported in this mode
E               raised from /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/typing/npydecl.py:114
E           In definition 1:
E               TypingError: ufunc 'isnan' using the loop 'l->?' not supported in this mode
E               raised from /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/typing/npydecl.py:114
E           This error is usually caused by passing an argument of a type that is unsupported by the named function.
E           [1] During: resolving callee type: Function(<ufunc 'isnan'>)
E           [2] During: typing of call at /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/targets/arraymath.py (2884)
E           
E           
E           File "../../../miniconda/envs/test-environment/lib/python3.7/site-packages/numba/targets/arraymath.py", line 2884:
E               def searchsorted_inner(a, v):
E                   <source elided>
E                   n = len(a)
E                   if np.isnan(v):
E                   ^
E           
E               raised from /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/typeinfer.py:861
E           In definition 1:
E               TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E           Invalid use of Function(<ufunc 'isnan'>) with argument(s) of type(s): (int64)
E            * parameterized
E           In definition 0:
E               TypingError: ufunc 'isnan' using the loop 'l->?' not supported in this mode
E               raised from /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/typing/npydecl.py:114
E           In definition 1:
E               TypingError: ufunc 'isnan' using the loop 'l->?' not supported in this mode
E               raised from /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/typing/npydecl.py:114
E           This error is usually caused by passing an argument of a type that is unsupported by the named function.
E           [1] During: resolving callee type: Function(<ufunc 'isnan'>)
E           [2] During: typing of call at /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/targets/arraymath.py (2884)
E           
E           
E           File "../../../miniconda/envs/test-environment/lib/python3.7/site-packages/numba/targets/arraymath.py", line 2884:
E               def searchsorted_inner(a, v):
E                   <source elided>
E                   n = len(a)
E                   if np.isnan(v):
E                   ^
E           
E               raised from /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/typeinfer.py:861
E           This error is usually caused by passing an argument of a type that is unsupported by the named function.
E           [1] During: resolving callee type: Function(<function _searchsorted.<locals>.searchsorted_inner at 0x7f37048f27b8>)
E           [2] During: typing of call at /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/targets/arraymath.py (2939)
E           
E           
E           File "../../../miniconda/envs/test-environment/lib/python3.7/site-packages/numba/targets/arraymath.py", line 2939:
E                   def searchsorted_impl(a, v, side='left'):
E                       return loop_impl(a, v)
E                       ^
E           
E               raised from /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/typeinfer.py:861
E           In definition 1:
E               ValueError: Invalid value given for 'side': unicode_type
E               raised from /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/numba/targets/arraymath.py:2917
E           This error is usually caused by passing an argument of a type that is unsupported by the named function.
E           [1] During: resolving callee type: Function(<function searchsorted at 0x7f37444bb840>)
E           [2] During: typing of call at /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/sparse/coo/indexing.py (451)
E           
E           
E           File "../../../miniconda/envs/test-environment/lib/python3.7/site-packages/sparse/coo/indexing.py", line 451:
E           def _get_mask_pairs(starts_old, stops_old, c, idx):  # pragma: no cover
E               <source elided>
E                   for p_match in range(idx[0], idx[1], idx[2]):
E                       start = np.searchsorted(c[starts_old[j]:stops_old[j]], p_match, side='left') + starts_old[j]
E                       ^
E           
E           [1] During: resolving callee type: type(CPUDispatcher(<function _get_mask_pairs at 0x7f37263d8268>))
E           [2] During: typing of call at /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/sparse/coo/indexing.py (387)
E           
E           
E           File "../../../miniconda/envs/test-environment/lib/python3.7/site-packages/sparse/coo/indexing.py", line 387:
E           def _compute_mask(coords, indices):  # pragma: no cover
E               <source elided>
E                   # Which would come out of indexing a single integer.
E                   starts, stops, n_matches = _get_mask_pairs(starts, stops, coords[i], indices[i])
E                   ^
E           
E           This is not usually a problem with Numba itself but instead often caused by
E           the use of unsupported features or an issue in resolving types.
E           
E           To see Python/NumPy features supported by the latest release of Numba visit:
E           http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
E           and
E           http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html
E           
E           For more information about typing errors and how to debug them visit:
E           http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile
E           
E           If you think your code should work with Numba, please report the error message
E           and traceback, along with a minimal reproducer at:
E           https://github.com/numba/numba/issues/new
../../../miniconda/envs/test-environment/lib/python3.7/site-packages/numba/six.py:658: TypingError
```

ref: https://travis-ci.org/dask/dask/jobs/515911552

jcrist commented:
The development builds often have periodic failures due to upstream issues that have nothing to do with us. As such, I'd prefer to either:

- Mark the development build as an allowed failure. This will still run the tests (so you can see if things broke when working on things that may affect upstream packages), but won't result in a red x on the PR.
- Make the development build an *optional* build, switched on some string in the pull request. This allows that build to be optionally turned on when changing functionality that may affect these features. We do this for the hdfs tests here: https://github.com/dask/dask/blob/631c5a4810565d08af674f0b8b537efaa0645295/.travis.yml#L71
  This may be less suited here, as the dev versions of dependencies may affect much of dask, but it's nice for features that rarely need to be tested.

hameerabbasi commented:
I've reported the issue upstream in Numba (numba/numba#3953)

pentschev commented:
I agree that we should find a middle-ground, like @jcrist suggested.

I would really like to get this PR merged soon, we're starting to stack different problems, preventing fixes from getting merged (like the FFT MKL workaround, which is by the way already stacked here, which isn't nice). So my proposal is the following:

* Revert my last commit and let Dask build against sparse release for Python 3.6 and sparse upstream for Python 3.7, to ensure we're properly testing this PR;
* Open another PR with the changes from my last commit so we can discuss and find a good solution.

After both items above are done, we can get this PR merged before it becomes a big snowball in an attempt to fix several partially-unrelated problems into one.

Do you agree with my proposal @jakirkham @mrocklin @jcrist @hameerabbasi ?

jakirkham commented:
Thanks @jcrist. This seems like a reasonable request to me. @pentschev, any thoughts on this?

Thanks for following up on that, @hameerabbasi. I see that issue is closed. What is the takeaway?

hameerabbasi commented:
It was a temporary break on Numba's end.

jakirkham commented:
Is there a known working version of Numba? Should we pin it to something?

hameerabbasi commented:
It was only in `master` AFAICT.

jakirkham commented:
Sorry I think I'm missing something. How would that cause the CI failure we saw? Are we getting a dev version of Numba somewhere that I missed?

pentschev commented:
@jakirkham I agree with @jcrist suggestions for that particular test build. In general I tend not to like this sort of test because it's likely to get ignored/forgotten, even when they're necessary. However, I'd still like to test against sparse upstream for every PR for a while, to ensure this keeps on working well.

Nevertheless, what do you think of my proposal @jakirkham from my last comment? I think we can and should think through the way we build, but I would prefer to have another PR for that as it's likely to be an issue that will have diverging opinions while holding off merging this PR.

hameerabbasi commented:
I just checked, apparently `UPSTREAM_DEV` doesn't update Numba and I did nothing in sparse to cause this. 🤔

pentschev commented:
`UPSTREAM_DEV` is probably updating Numba as a dependency to some other package. Passing builds are using Numba 0.41.0, whereas the failing `UPSTREAM_DEV` build is using Numba 0.43.1.

hameerabbasi commented:
Anyway, I isolated the issue (as it was essentially a compilation issue, which is easy to isolate) and made sure it's not there in Numba `master`.

pentschev commented:
> Anyway, I isolated the issue (as it was essentially a compilation issue, which is easy to isolate) and made sure it's not there in Numba `master`.

Thanks so much for doing that @hameerabbasi.

One alternative we have now is to try using Numba master for `UPSTREAM_DEV` as well, but I think this may make us go in the direction of digging into further problems.

pentschev commented:
Since there were no objections to my proposal, I reverted the last commit so that in this PR we can focus on the fixes that we aim to have while moving the discussion on how to move forward with the test of upstream libraries to #4696.

Please, let me know if there are any other suggestions for the fixes here, otherwise, I think this was already good for merging a couple of weeks ago.

jakirkham commented:
Thanks @pentschev!
